### PR TITLE
Use the new trait solver.

### PR DESCRIPTION
**Pull request by jhjourdan**
We will nee too do that anyway, and this may help for #1232.

The main issue I encountered is that the new trait solver does not allow ambiguity between a user impl and an impl coming from ParamEnv when searching for an instance. This is fixed by removing all such ambiguities in our code base. At some point, we will need to figure out a way to report a pleasant error message when finding trait method instances fails because of such an ambiguity.

### CODE REVIEW

File: creusot-contracts/src/std/ops.rs [hunk 1]
```diff
@@ -16,7 +18,7 @@ pub trait FnOnceExt<Args: Tuple>: FnOnce<Args> {
 /// `FnMutExt` is an extension trait for the `FnMut` trait, used for
 /// adding a specification to closures. It should not be used directly.
 #[rustc_diagnostic_item = "fn_mut_spec"]
-pub trait FnMutExt<Args: Tuple>: FnMut<Args> + FnOnceExt<Args> {
```
Comment (xldenis) [inline creusot-contracts/src/std/ops.rs:19]: 🎉

File: creusot-contracts/src/std/ops.rs [hunk 2]
```diff
@@ -5,7 +5,9 @@ pub use ::std::ops::*;
 /// `FnOnceExt` is an extension trait for the `FnOnce` trait, used for
 /// adding a specification to closures. It should not be used directly.
 #[rustc_diagnostic_item = "fn_once_spec"]
-pub trait FnOnceExt<Args: Tuple>: FnOnce<Args> {
+pub trait FnOnceExt<Args: Tuple> {
+    type Output;
```
Comment (xldenis) [inline creusot-contracts/src/std/ops.rs:9]: Interesting, you needed to do this?

Comment (jhjourdan) [reply creusot-contracts/src/std/ops.rs:9]: This creates an ambiguity in trait resolution, which the new solver does not like.

**Comment by jhjourdan** [comment]
Oh, and by the way: this PR magically fixes our problem with `PartialEq::eq`.

**Comment by xldenis** [comment]
> Oh, and by the way: this PR magically fixes our problem with PartialEq::eq.

Magical fixes are the best kinds of fixes

**Pull request merged**